### PR TITLE
Add bce_loss op

### PR DESCRIPTION
**Pull request by ceci3**
Add bce_loss op, `paddle.nn.loss.BCELoss(weight=None, reduction='mean')`.

base format is `Out = -1 * (label * log(input) + (1 - label) * log(1 - input))`

examples:
```python
            import paddle.fluid as fluid
            import numpy as np
            input = fluid.data(name="input", shape=[3, 1])
            label = fluid.data(name="label", shape=[3, 1])
            output = fluid.layers.bce_loss(input,label)
            place = fluid.CPUPlace()
            exe = fluid.Executor(place)
            exe.run(fluid.default_startup_program())

            input_data = np.array([0.5, 0.6, 0.7]).astype("float32")
            label_data = np.array([1.0, 0.0, 1.0]).astype("float32")
            output_data = exe.run(fluid.default_main_program(),
                    feed={"input":input_data, "label":label_data},
                    fetch_list=[output],
                    return_numpy=True)

            print(output_data)  # [array([0.65537095], dtype=float32)]
```

### CODE REVIEW

File: python/paddle/nn/layer/loss.py [hunk 1]
```diff
@@ -109,3 +109,110 @@ def forward(self, input, label):
...
+
+    def forward(self, input, label):
+        dtype = self._helper.input_dtype(input)
+
+        fluid.data_feeder.check_variable_and_dtype(
+            input, 'input', ['float32', 'float64'], 'bce_loss')
+        fluid.data_feeder.check_variable_and_dtype(
+            label, 'label', ['float32', 'float64'], 'bce_loss')
+
+        if self.reduction not in ['sum', 'mean', 'none']:
```
Comment (LielinJiang) [inline python/paddle/nn/layer/loss.py]: move to __init__()? because `reduction` isn't a variable

File: paddle/fluid/operators/bce_loss_op.cu [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+#include "paddle/fluid/operators/math.h"
+#include "paddle/fluid/platform/cuda_primitives.h"
+#include "paddle/fluid/platform/hostdevice.h"
+
+namespace paddle {
+namespace operators {
+
+using Tensor = framework::Tensor;
+
+static constexpr int kNumCUDAThreads = 512;
```
Comment (LielinJiang) [inline paddle/fluid/operators/bce_loss_op.cu]: Use GpuLaunchConfig to define blocks and threads may get better performance, you can reference this pr:https://github.com/PaddlePaddle/Paddle/pull/22436

File: python/paddle/nn/layer/loss.py (hunk 1 above)
Comment (ceci3) [reply python/paddle/nn/layer/loss.py]: done

File: paddle/fluid/operators/bce_loss_op.cu (hunk 2 above)
Comment (ceci3) [reply paddle/fluid/operators/bce_loss_op.cu]: done

**Review by jerrywgz — DISMISSED** [review]
LGTM

**Review by jerrywgz — DISMISSED** [review]

File: paddle/fluid/operators/bce_loss_op.cc [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+using framework::Tensor;
+
+class BCELossOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+  void InferShape(framework::InferShapeContext* ctx) const override {
+    PADDLE_ENFORCE_EQ(
+        ctx->HasInput("X"), true,
+        platform::errors::InvalidArgument("Input(X) should be not null."));
```
Comment (heavengate) [inline paddle/fluid/operators/bce_loss_op.cc:32]: InvalidArgument -> NotFound

File: paddle/fluid/operators/bce_loss_op.cc [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+  using framework::SingleGradOpMaker<T>::SingleGradOpMaker;
+
+ protected:
+  void Apply(GradOpPtr<T> op) const override {
+    op->SetType("bce_loss_grad");
+    op->SetInput("X", this->Input("X"));
+    op->SetInput("Label", this->Input("Label"));
+    op->SetInput(framework::GradVarName("Out"), this->OutputGrad("Out"));
+    op->SetOutput(framework::GradVarName("X"), this->InputGrad("X"));
+    //    op->SetAttrMap(this->Attrs());
```
Comment (heavengate) [inline paddle/fluid/operators/bce_loss_op.cc:155]: uncomment or delete

File: python/paddle/fluid/tests/unittests/test_bce_loss.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        self.outputs = {'Out': output_np}
+
+    def test_check_output(self):
+        self.check_output()
+
+    def test_check_grad(self):
+        self.check_grad(['X'], 'Out')
+
+    def init_test_case(self):
+        self.shape = [10, 10]
```
Comment (heavengate) [inline python/paddle/fluid/tests/unittests/test_bce_loss.py:131]: 这里最好再给几个3-D 4-D的testcase

Comment (ceci3) [reply python/paddle/fluid/tests/unittests/test_bce_loss.py:131]: done in https://github.com/PaddlePaddle/Paddle/pull/23746

File: paddle/fluid/operators/bce_loss_op.cc (hunk 4 above)
Comment (ceci3) [reply paddle/fluid/operators/bce_loss_op.cc:155]: done in #23746

File: paddle/fluid/operators/bce_loss_op.cc (hunk 3 above)
Comment (ceci3) [reply paddle/fluid/operators/bce_loss_op.cc:32]: done in #23746

**Review by heavengate — APPROVED** [review]

**Pull request merged**